dvc/parsing/__init__.py

```diff
@@ -453,7 +458,99 @@ def _each_iter(self, key: str) -> "DictStrAny":
             except ContextError as exc:
                 format_and_raise(exc, f"stage '{generated}'", self.relpath)
 
-            # let mypy know that this state is unreachable as format_and_raise
-            # does not return at all (it's not able to understand it for some
-            # reason)
-            raise AssertionError("unreachable")
+
+class MatrixDefinition:
```

Comment:
Minor: looks like several of the methods here are duplicated with ForeachDefinition

Reply:
They share common APIs, basically it's a protocol, but implementation is very different.
It's too early to DRY the code.